File: quantum-sensing-jl/states.jl
# State parametrization: Fock superpositions and squeezed/coherent superpositions.

# ── Wavefunction utilities ────────────────────────────────────────────────

export harmonic_oscillator_wavefunction, wavefunction_to_quantumobject

"""
    harmonic_oscillator_wavefunction(n, q)

n-th harmonic oscillator wavefunction φ_n(q) evaluated at positions `q`.
"""
function harmonic_oscillator_wavefunction(n::Int, q::AbstractVector{<:Real})
    # Use the stable recurrence on the full wavefunctions (not raw Hermite polynomials)
    # φ_0(q) = π^{-1/4} exp(-q²/2)
    # φ_{n+1}(q) = √(2/(n+1)) q φ_n(q) - √(n/(n+1)) φ_{n-1}(q)
    gauss = exp.(-q .^ 2 ./ 2.0)
    φ_prev = (π ^ (-0.25)) .* gauss  # φ_0

    n == 0 && return φ_prev

    φ_curr = sqrt(2.0) .* q .* φ_prev  # φ_1

    for k in 1:(n-1)
        φ_next = sqrt(2.0 / (k + 1)) .* q .* φ_curr .- sqrt(k / (k + 1)) .* φ_prev
        φ_prev = φ_curr
        φ_curr = φ_next
    end
    return φ_curr
end

"""
    _ho_basis_matrix(q, N_basis) -> Matrix{Float64}

Build the N_basis × length(q) matrix of harmonic oscillator wavefunctions
using a single-pass three-term recurrence.  Row n+1 holds φ_n(q).
"""
function _ho_basis_matrix(q::AbstractVector{<:Real}, N_basis::Int)
    Nq = length(q)
    Φ = Matrix{Float64}(undef, N_basis, Nq)

    gauss = exp.(-q .^ 2 ./ 2.0)
    Φ[1, :] .= (π ^ (-0.25)) .* gauss  # φ_0

    if N_basis >= 2
        Φ[2, :] .= sqrt(2.0) .* q .* Φ[1, :]  # φ_1
    end

    for n in 2:(N_basis - 1)
        @views Φ[n+1, :] .= sqrt(2.0 / n) .* q .* Φ[n, :] .- sqrt((n - 1) / n) .* Φ[n-1, :]
    end

    return Φ
end

"""
    _simpson_weights(q) -> Vector{Float64}

Composite Simpson's rule weights for non-uniform spacing.
Falls back to trapezoidal for odd-length input.
"""
function _simpson_weights(q::AbstractVector{<:Real})
    N = length(q)
    w = zeros(Float64, N)

    if N < 3 || isodd(N)
        # Trapezoidal fallback
        dq = diff(q)
        w[1] = dq[1] / 2
        for i in 2:(N-1)
            w[i] = (dq[i-1] + dq[i]) / 2
        end
        w[N] = dq[end] / 2
        return w
    end

    # Composite Simpson on pairs of intervals
    for i in 1:2:(N-2)
        h = q[i+2] - q[i]
        h1 = q[i+1] - q[i]
        h2 = q[i+2] - q[i+1]
        w[i]   += h * (2*h1 - h2) / (6*h1)
        w[i+1] += h^3 / (6*h1*h2)
        w[i+2] += h * (2*h2 - h1) / (6*h2)
    end

    return w
end

"""
    wavefunction_to_quantumobject(q, ψ; N_basis=25)

Convert a position-space wavefunction ψ(q) to a QuantumToolbox ket in the Fock basis.

Computes all N_basis overlaps ⟨n|ψ⟩ = ∫ φ_n(q) ψ(q) dq in a single pass
using the HO three-term recurrence and Simpson's rule integration.
"""
function wavefunction_to_quantumobject(q::AbstractVector{<:Real},
                                        ψ::AbstractVector{<:Number};
                                        N_basis::Int=25)
    w = _simpson_weights(q)

    # Normalize input: ∫|ψ|² dq = 1
    norm2 = dot(w, abs2.(ψ))
    ψ_norm = ψ ./ sqrt(norm2)

    # Weighted wavefunction: w .* ψ_norm
    wψ = w .* ψ_norm

    # Build basis matrix (N_basis × Nq) and compute all overlaps at once
    Φ = _ho_basis_matrix(q, N_basis)
    coeffs = Φ * wψ  # matrix-vector product: (N_basis × Nq) × (Nq,) → (N_basis,)

    state = QuantumObject(ComplexF64.(coeffs), dims=(N_basis,))
    return normalize(state)
end

# ── Constants ─────────────────────────────────────────────────────────────

export STATE_TYPES, LOSS_CONFIGS

const STATE_TYPES  = [:sqz_vac, :sqz_coh, :coherent_sup]
const LOSS_CONFIGS = [:loss_in, :loss_ch, :loss_out]

# ══════════════════════════════════════════════════════════════════════════
# Fock sparse superpositions:  |ψ⟩ = N Σ_k c_k |n_k⟩
#   3 reals per term: [n_float, c_re, c_im]
# ══════════════════════════════════════════════════════════════════════════

export fock_params_per_state, fock_unpack_params, fock_pack_params, fock_get_bounds,
       fock_create_state, fock_unpack_coeff_params, fock_pack_coeff_params, fock_get_coeff_bounds

fock_params_per_state() = 3

function fock_unpack_params(x::AbstractVector, n_sup::Int)
    @assert length(x) == 3 * n_sup "Expected $(3*n_sup) params, got $(length(x))"
    n_floats = Vector{Float64}(undef, n_sup)
    c_vals   = Vector{ComplexF64}(undef, n_sup)
    for k in 1:n_sup
        base       = (k - 1) * 3
        n_floats[k] = x[base + 1]
        c_vals[k]   = Complex(x[base + 2], x[base + 3])
    end
    return n_floats, c_vals
end

function fock_pack_params(n_floats::AbstractVector, c_vals::AbstractVector)::Vector{Float64}
    n_sup = length(n_floats)
    x = Vector{Float64}(undef, 3 * n_sup)
    for k in 1:n_sup
        base       = (k - 1) * 3
        x[base + 1] = n_floats[k]
        x[base + 2] = real(c_vals[k])
        x[base + 3] = imag(c_vals[k])
    end
    return x
end

function fock_get_bounds(n_sup::Int, N_basis::Int)
    lb = Float64[]
    ub = Float64[]
    for _ in 1:n_sup
        append!(lb, [0.0,  -1.0, -1.0])
        append!(ub, [Float64(N_basis - 1),  1.0,  1.0])
    end
    return lb, ub
end

function fock_unpack_coeff_params(x_coeff::AbstractVector, n_sup::Int)
    @assert length(x_coeff) == 2 * n_sup
    c_vals = Vector{ComplexF64}(undef, n_sup)
    for k in 1:n_sup
        base     = (k - 1) * 2
        c_vals[k] = Complex(x_coeff[base + 1], x_coeff[base + 2])
    end
    return c_vals
end

function fock_pack_coeff_params(c_vals::AbstractVector)::Vector{Float64}
    n_sup = length(c_vals)
    x = Vector{Float64}(undef, 2 * n_sup)
    for k in 1:n_sup
        base     = (k - 1) * 2
        x[base + 1] = real(c_vals[k])
        x[base + 2] = imag(c_vals[k])
    end
    return x
end

function fock_get_coeff_bounds(n_sup::Int)
    lb = fill(-1.0, 2 * n_sup)
    ub = fill( 1.0, 2 * n_sup)
    return lb, ub
end

"""
    fock_create_state(n_floats, c_vals, N_basis) -> ket

Builds a normalised Fock superposition |ψ⟩ = N Σ_k c_k |round(n_k)⟩.
"""
function fock_create_state(n_floats::AbstractVector, c_vals::AbstractVector, N_basis::Int)
    ψ_data = zeros(ComplexF64, N_basis)
    for k in eachindex(c_vals)
        abs(c_vals[k]) < 1e-12 && continue
        n_k = clamp(round(Int, n_floats[k]), 0, N_basis - 1)
        ψ_data[n_k + 1] += c_vals[k]
    end
    if norm(ψ_data) < 1e-10
        ψ_data[1] = 1.0
    end
    return normalize(QuantumObject(ψ_data, dims=(N_basis,)))
end

# ══════════════════════════════════════════════════════════════════════════
# Squeezed/coherent superpositions:  |ψ⟩ = N Σ_k c_k D(α_k)S(r_k)|0⟩
#   :sqz_vac      — 4 reals: [c_re, c_im, r_re, r_im]
#   :sqz_coh      — 6 reals: [c_re, c_im, α_re, α_im, r_re, r_im]
#   :coherent_sup — 4 reals: [c_re, c_im, α_re, α_im]
# ══════════════════════════════════════════════════════════════════════════

export sqz_params_per_state, sqz_unpack_params, sqz_pack_params, sqz_get_bounds,
       sqz_create_state, embed_params_into_sqz_coh

function sqz_params_per_state(state_type::Symbol)::Int
    state_type == :sqz_vac     && return 4
    state_type == :sqz_coh     && return 6
    state_type == :coherent_sup && return 4
    error("Unknown state_type: $state_type. Choose from $STATE_TYPES")
end

function sqz_unpack_params(state_type::Symbol, x::AbstractVector, n_sups::Int)
    p = sqz_params_per_state(state_type)
    @assert length(x) == p * n_sups "Expected $(p*n_sups) params, got $(length(x))"

    c_vals     = Vector{ComplexF64}(undef, n_sups)
    alpha_vals = zeros(ComplexF64, n_sups)
    r_vals     = zeros(ComplexF64, n_sups)

    for k in 1:n_sups
        base = (k - 1) * p
        if state_type == :sqz_vac
            c_vals[k] = Complex(x[base+1], x[base+2])
            r_vals[k] = Complex(x[base+3], x[base+4])
        elseif state_type == :sqz_coh
            c_vals[k]     = Complex(x[base+1], x[base+2])
            alpha_vals[k] = Complex(x[base+3], x[base+4])
            r_vals[k]     = Complex(x[base+5], x[base+6])
        elseif state_type == :coherent_sup
            c_vals[k]     = Complex(x[base+1], x[base+2])
            alpha_vals[k] = Complex(x[base+3], x[base+4])
        end
    end
    return c_vals, alpha_vals, r_vals
end

function sqz_pack_params(state_type::Symbol,
                         c_vals::AbstractVector,
                         alpha_vals::AbstractVector,
                         r_vals::AbstractVector)::Vector{Float64}
    n_sups = length(c_vals)
    p = sqz_params_per_state(state_type)
    x = Vector{Float64}(undef, p * n_sups)

    for k in 1:n_sups
        base = (k - 1) * p
        if state_type == :sqz_vac
            x[base+1] = real(c_vals[k]);     x[base+2] = imag(c_vals[k])
            x[base+3] = real(r_vals[k]);     x[base+4] = imag(r_vals[k])
        elseif state_type == :sqz_coh
            x[base+1] = real(c_vals[k]);     x[base+2] = imag(c_vals[k])
            x[base+3] = real(alpha_vals[k]); x[base+4] = imag(alpha_vals[k])
            x[base+5] = real(r_vals[k]);     x[base+6] = imag(r_vals[k])
        elseif state_type == :coherent_sup
            x[base+1] = real(c_vals[k]);     x[base+2] = imag(c_vals[k])
            x[base+3] = real(alpha_vals[k]); x[base+4] = imag(alpha_vals[k])
        end
    end
    return x
end

function sqz_get_bounds(state_type::Symbol, n_sups::Int, N_target::Real)
    c_lim     = 1.0
    alpha_lim = sqrt(max(N_target, 1.0))
    r_lim     = max(asinh(2 * N_target), 2.0)

    lb = Float64[]
    ub = Float64[]

    for _ in 1:n_sups
        if state_type == :sqz_vac
            append!(lb, [-c_lim, -c_lim, -r_lim, -r_lim])
            append!(ub, [ c_lim,  c_lim,  r_lim,  r_lim])
        elseif state_type == :sqz_coh
            append!(lb, [-c_lim, -c_lim, -alpha_lim, -alpha_lim, -r_lim, -r_lim])
            append!(ub, [ c_lim,  c_lim,  alpha_lim,  alpha_lim,  r_lim,  r_lim])
        elseif state_type == :coherent_sup
            append!(lb, [-c_lim, -c_lim, -alpha_lim, -alpha_lim])
            append!(ub, [ c_lim,  c_lim,  alpha_lim,  alpha_lim])
        end
    end
    return lb, ub
end

function _create_component(α::Number, r::Number, N_basis::Int)
    ψ = fock(N_basis, 0)
    abs(r) > 1e-10 && (ψ = squeeze(N_basis, r) * ψ)
    abs(α) > 1e-10 && (ψ = displace(N_basis, α) * ψ)
    return ψ
end

"""
    sqz_create_state(c_vals, alpha_vals, r_vals, N_basis) -> ket

Builds a normalised superposition  |ψ⟩ = N Σ_k c_k D(α_k)S(r_k)|0⟩.
"""
function sqz_create_state(c_vals::AbstractVector,
                          alpha_vals::AbstractVector,
                          r_vals::AbstractVector,
                          N_basis::Int)
    n = length(c_vals)
    ψ_super = zeros(ComplexF64, N_basis)

    for k in 1:n
        abs(c_vals[k]) < 1e-12 && continue
        ψ_k = _create_component(alpha_vals[k], r_vals[k], N_basis)
        ψ_super .+= c_vals[k] .* ψ_k.data
    end

    return normalize(QuantumObject(ψ_super, dims=(N_basis,)))
end

"""
    embed_params_into_sqz_coh(source_type, source_params, n_sups) -> Vector{Float64}

Converts a parameter vector from :sqz_vac or :coherent_sup into the :sqz_coh layout.
"""
function embed_params_into_sqz_coh(source_type::Symbol,
                                    source_params::AbstractVector,
                                    n_sups::Int)::Vector{Float64}
    p_src = sqz_params_per_state(source_type)
    @assert length(source_params) == p_src * n_sups

    x_cat = zeros(Float64, 6 * n_sups)

    for k in 1:n_sups
        src_base  = (k - 1) * p_src
        cat_base  = (k - 1) * 6

        c_re, c_im = source_params[src_base+1], source_params[src_base+2]
        x_cat[cat_base+1] = c_re
        x_cat[cat_base+2] = c_im

        if source_type == :sqz_vac
            x_cat[cat_base+5] = source_params[src_base+3]
            x_cat[cat_base+6] = source_params[src_base+4]
        elseif source_type == :coherent_sup
            x_cat[cat_base+3] = source_params[src_base+3]
            x_cat[cat_base+4] = source_params[src_base+4]
        end
    end

    return x_cat
end

File: 2026-09-07/quantum_sensing/states.py
"""
Unified helper for loading quantum state optimization results and reconstructing
QuTiP quantum states for fock_sup, sqz_vac, and sqz_coh state types.

Data directory must be configured before using load functions:
    import quantum_sensing
    quantum_sensing.set_data_dir("/path/to/consolidated_data")

Or via environment variable:
    export QUANTUM_SENSING_DATA_DIR=/path/to/consolidated_data
"""

import os
from pathlib import Path
import numpy as np
import pandas as pd
import h5py
import qutip as qt

# ---- Data directory configuration ----------------------------------------

_DATA_DIR = None


def set_data_dir(path):
    """Set the path to the consolidated_data directory."""
    global _DATA_DIR
    _DATA_DIR = Path(path)


def _get_data_dir():
    global _DATA_DIR
    if _DATA_DIR is not None:
        return _DATA_DIR
    env = os.environ.get("QUANTUM_SENSING_DATA_DIR")
    if env:
        return Path(env)
    raise RuntimeError(
        "Data directory not configured. Either:\n"
        "  - Call quantum_sensing.set_data_dir('/path/to/consolidated_data')\n"
        "  - Set QUANTUM_SENSING_DATA_DIR environment variable"
    )


# ---- Style constants (import * gives these to notebooks) -----------------

STATE_COLORS = {
    "fock_sup": "#1f77b4",
    "sqz_vac":  "#d62728",
    "sqz_coh":  "#2ca02c",
}
STATE_LABELS = {
    "fock_sup": "Fock Sup.",
    "sqz_vac":  "Sqz. Vac. Sup.",
    "sqz_coh":  "Sqz. Coh. Sup.",
}
STATE_MARKERS = {
    "fock_sup": "o",
    "sqz_vac":  "s",
    "sqz_coh":  "^",
}

# ---- Data loading --------------------------------------------------------

def load_results_df() -> pd.DataFrame:
    """Load master_results.csv, excluding rows marked valid=false."""
    csv_path = _get_data_dir() / "master_results.csv"
    if not csv_path.exists():
        raise FileNotFoundError(
            f"master_results.csv not found at {csv_path}\n"
            "Run rebuild_master_csv() from Julia to generate it."
        )
    df = pd.read_csv(csv_path, dtype={"valid": str})
    return df[df["valid"].fillna("").str.lower() != "false"].reset_index(drop=True)


def load_state_params(state_type: str, loss_config: str,
                      eta: float, pn: float) -> list[dict]:
    """
    Load all result dicts for one (state_type, loss_config, eta, pn) slice
    from states.h5.

    Returns a list of dicts with keys:
        state_type, loss_config, eta, pn,
        N_target, n_sups, N_basis, qfi, N_actual,
        c_vals, alpha_vals, r_vals, n_vals   (complex/int numpy arrays)
    """
    h5_path = _get_data_dir() / "states.h5"
    if not h5_path.exists():
        raise FileNotFoundError(
            f"states.h5 not found at {h5_path}\n"
            "Run: julia --project=julia_code consolidated_data/scripts/rebuild_all.jl"
        )

    eta_str = _fmt(eta)
    pn_str  = _fmt(pn)
    group_key  = f"{state_type}_{loss_config}"
    eta_pn_key = f"eta{eta_str}_pn{pn_str}"

    results = []
    with h5py.File(h5_path, "r") as h5f:
        if group_key not in h5f:
            return results
        grp = h5f[group_key]
        if eta_pn_key not in grp:
            return results
        eta_grp = grp[eta_pn_key]

        for entry_key, entry in eta_grp.items():
            attrs = dict(entry.attrs)
            c_re = entry["c_vals_re"][()]
            c_im = entry["c_vals_im"][()]
            a_re = entry["alpha_re"][()]
            a_im = entry["alpha_im"][()]
            r_re = entry["r_re"][()]
            r_im = entry["r_im"][()]
            nv   = entry["n_vals"][()].astype(int)

            results.append({
                "state_type":  state_type,
                "loss_config": loss_config,
                "eta":         float(eta),
                "pn":          float(pn),
                "N_target":    float(attrs.get("N_target", np.nan)),
                "n_sups":      int(attrs.get("n_sups", 0)),
                "N_basis":     int(attrs.get("N_basis", 0)),
                "qfi":         float(attrs.get("qfi", np.nan)),
                "N_actual":    float(attrs.get("N_actual", np.nan)),
                "c_vals":      c_re + 1j * c_im,
                "alpha_vals":  a_re + 1j * a_im,
                "r_vals":      r_re + 1j * r_im,
                "n_vals":      nv,
            })

    return results


def list_available_eta_pn(state_type: str, loss_config: str) -> list[str]:
    """List available eta_pn keys in states.h5 for a given (state_type, loss_config)."""
    h5_path = _get_data_dir() / "states.h5"
    if not h5_path.exists():
        return []
    group_key = f"{state_type}_{loss_config}"
    with h5py.File(h5_path, "r") as h5f:
        if group_key not in h5f:
            return []
        return sorted(h5f[group_key].keys())

# ---- Trend queries (DataFrame ops) --------------------------------------

def _filter(df: pd.DataFrame, state_type: str, loss_config: str,
            eta: float = None, pn: float = None,
            N_target: float = None, n_sups: int = None,
            tol: float = 1e-6) -> pd.DataFrame:
    mask = (df["state_type"] == state_type) & (df["loss_config"] == loss_config)
    if eta is not None:
        mask &= (df["eta"] - eta).abs() < tol
    if pn is not None:
        mask &= (df["pn"] - pn).abs() < tol
    if N_target is not None:
        mask &= (df["N_target"] - N_target).abs() < 0.01
    if n_sups is not None:
        mask &= (df["n_sups"] == n_sups)
    return df[mask]


def get_best_qfi_envelope(df: pd.DataFrame, state_type: str, loss_config: str,
                          eta: float, pn: float):
    """Max QFI over n_sups per N_target. Returns (N_arr, qfi_arr) or None."""
    sub = _filter(df, state_type, loss_config, eta=eta, pn=pn)
    sub = sub[sub["qfi"] > 0].dropna(subset=["qfi"])
    if sub.empty:
        return None
    best = sub.groupby("N_target")["qfi"].max().reset_index().sort_values("N_target")
    return best["N_target"].to_numpy(), best["qfi"].to_numpy()


def get_nsup_trend(df: pd.DataFrame, state_type: str, loss_config: str,
                   eta: float, pn: float, N_target: float):
    """QFI vs n_sups at fixed (eta, pn, N_target). Returns (ns_arr, qfi_arr) or None."""
    sub = _filter(df, state_type, loss_config, eta=eta, pn=pn, N_target=N_target)
    sub = sub[sub["qfi"] > 0].dropna(subset=["qfi"])
    if sub.empty:
        return None
    sub = sub.sort_values("n_sups")
    return sub["n_sups"].to_numpy(), sub["qfi"].to_numpy()


def get_eta_trend(df: pd.DataFrame, state_type: str, loss_config: str,
                  pn: float, N_target: float):
    """Best QFI vs eta at fixed (pn, N_target). Returns (eta_arr, qfi_arr) or None."""
    sub = _filter(df, state_type, loss_config, pn=pn, N_target=N_target)
    sub = sub[sub["qfi"] > 0].dropna(subset=["qfi"])
    if sub.empty:
        return None
    best = sub.groupby("eta")["qfi"].max().reset_index().sort_values("eta")
    return best["eta"].to_numpy(), best["qfi"].to_numpy()


def get_pn_trend(df: pd.DataFrame, state_type: str, loss_config: str,
                 eta: float, N_target: float):
    """Best QFI vs pn at fixed (eta, N_target). Returns (pn_arr, qfi_arr) or None."""
    sub = _filter(df, state_type, loss_config, eta=eta, N_target=N_target)
    sub = sub[sub["qfi"] > 0].dropna(subset=["qfi"])
    if sub.empty:
        return None
    best = sub.groupby("pn")["qfi"].max().reset_index().sort_values("pn")
    return best["pn"].to_numpy(), best["qfi"].to_numpy()

# ---- Wavefunction utilities ----------------------------------------------

def harmonic_oscillator_wavefunction(n, q):
    """
    n-th harmonic oscillator wavefunction phi_n(q) evaluated at positions q.

    Uses the stable recurrence on full wavefunctions (not raw Hermite polynomials):
        phi_0(q) = pi^{-1/4} exp(-q^2/2)
        phi_{n+1}(q) = sqrt(2/(n+1)) q phi_n(q) - sqrt(n/(n+1)) phi_{n-1}(q)
    """
    q = np.asarray(q, dtype=float)
    gauss = np.exp(-q**2 / 2.0)
    phi_prev = (np.pi ** (-0.25)) * gauss  # phi_0

    if n == 0:
        return phi_prev

    phi_curr = np.sqrt(2.0) * q * phi_prev  # phi_1

    for k in range(1, n):
        phi_next = np.sqrt(2.0 / (k + 1)) * q * phi_curr - np.sqrt(k / (k + 1)) * phi_prev
        phi_prev = phi_curr
        phi_curr = phi_next

    return phi_curr


def _ho_basis_matrix(q, N_basis):
    """
    Build (N_basis, len(q)) matrix of HO wavefunctions via single-pass recurrence.
    Row n holds phi_n(q).
    """
    q = np.asarray(q, dtype=float)
    Nq = len(q)
    Phi = np.empty((N_basis, Nq), dtype=float)

    gauss = np.exp(-q**2 / 2.0)
    Phi[0] = (np.pi ** (-0.25)) * gauss

    if N_basis >= 2:
        Phi[1] = np.sqrt(2.0) * q * Phi[0]

    for n in range(2, N_basis):
        Phi[n] = np.sqrt(2.0 / n) * q * Phi[n-1] - np.sqrt((n-1) / n) * Phi[n-2]

    return Phi


def _simpson_weights(q):
    """Composite Simpson's rule weights. Falls back to trapezoidal for odd-length input."""
    q = np.asarray(q, dtype=float)
    N = len(q)
    w = np.zeros(N)

    if N < 3 or N % 2 == 0:
        # Trapezoidal fallback (N is even means odd number of intervals)
        dq = np.diff(q)
        w[0] = dq[0] / 2
        w[1:-1] = (dq[:-1] + dq[1:]) / 2
        w[-1] = dq[-1] / 2
        return w

    # Composite Simpson on pairs of intervals
    for i in range(0, N - 2, 2):
        h = q[i+2] - q[i]
        h1 = q[i+1] - q[i]
        h2 = q[i+2] - q[i+1]
        w[i]   += h * (2*h1 - h2) / (6*h1)
        w[i+1] += h**3 / (6*h1*h2)
        w[i+2] += h * (2*h2 - h1) / (6*h2)

    return w


def wavefunction_to_qobj(q, psi, N_basis=25):
    """
    Convert a position-space wavefunction to a QuTiP Qobj in the Fock basis.

    Computes all N_basis overlaps <n|psi> in a single pass using the HO
    three-term recurrence and Simpson's rule integration.

    Args:
        q: 1D array of position coordinates.
        psi: 1D array of wavefunction values (can be complex).
        N_basis: Hilbert space dimension (Fock basis cutoff).

    Returns:
        qt.Qobj: Normalized ket in the Fock basis.
    """
    q = np.asarray(q, dtype=float)
    psi = np.asarray(psi, dtype=complex)

    w = _simpson_weights(q)

    # Normalize input
    norm2 = np.sum(w * np.abs(psi)**2)
    psi_norm = psi / np.sqrt(norm2)

    # Weighted wavefunction
    w_psi = w * psi_norm

    # Build basis matrix and compute all overlaps via matrix-vector product
    Phi = _ho_basis_matrix(q, N_basis)
    coeffs = Phi @ w_psi  # (N_basis, Nq) @ (Nq,) -> (N_basis,)

    state_qobj = qt.Qobj(coeffs, dims=[[N_basis], [1]])
    return state_qobj.unit()


# ---- State reconstruction ------------------------------------------------

def reconstruct_fock_sup(n_vals, c_vals, N_basis: int) -> qt.Qobj:
    """Build |psi> = sum_k c_k |n_k> and normalise."""
    psi = qt.Qobj(np.zeros(N_basis, dtype=complex))
    for n, c in zip(n_vals, c_vals):
        n = int(n)
        if 0 <= n < N_basis:
            psi += c * qt.basis(N_basis, n)
    return psi.unit()


def reconstruct_sqz_vac(c_vals, r_vals, N_basis: int) -> qt.Qobj:
    """Build |psi> = sum_k c_k S(r_k)|0> and normalise."""
    vac = qt.basis(N_basis, 0)
    psi = qt.Qobj(np.zeros(N_basis, dtype=complex))
    for c, r in zip(c_vals, r_vals):
        if abs(c) < 1e-14:
            continue
        sq = qt.squeeze(N_basis, complex(r)) * vac if abs(r) > 1e-12 else vac
        psi += c * sq
    return psi.unit()


def reconstruct_sqz_coh(c_vals, alpha_vals, r_vals, N_basis: int) -> qt.Qobj:
    """Build |psi> = sum_k c_k D(alpha_k) S(r_k)|0> and normalise."""
    vac = qt.basis(N_basis, 0)
    psi = qt.Qobj(np.zeros(N_basis, dtype=complex))
    for c, alpha, r in zip(c_vals, alpha_vals, r_vals):
        if abs(c) < 1e-14:
            continue
        ket = vac
        if abs(r) > 1e-12:
            ket = qt.squeeze(N_basis, complex(r)) * ket
        if abs(alpha) > 1e-12:
            ket = qt.displace(N_basis, complex(alpha)) * ket
        psi += c * ket
    return psi.unit()


def reconstruct_state(result: dict, N_basis: int = None) -> qt.Qobj:
    """Dispatch state reconstruction based on result['state_type'].

    Parameters
    ----------
    result  : dict returned by load_state_params()
    N_basis : override the stored Hilbert-space dimension (e.g. to up-sample
              or down-sample). Defaults to result['N_basis'].
    """
    st     = result["state_type"]
    N      = N_basis if N_basis is not None else result["N_basis"]
    c_vals = result["c_vals"]

    if st == "fock_sup":
        return reconstruct_fock_sup(result["n_vals"], c_vals, N)
    elif st == "sqz_vac":
        return reconstruct_sqz_vac(c_vals, result["r_vals"], N)
    elif st in ("sqz_coh", "sqz_cat"):
        return reconstruct_sqz_coh(c_vals, result["alpha_vals"], result["r_vals"], N)
    else:
        raise ValueError(f"Unknown state_type: {st!r}")


def get_wigner(result: dict, xrange: tuple = (-5.0, 5.0), pts: int = 200,
               N_basis: int = None):
    """
    Reconstruct state from result dict and compute Wigner function.

    Parameters
    ----------
    N_basis : override Hilbert-space dimension (passed to reconstruct_state).

    Returns (xvec, W) where W has shape (pts, pts).
    """
    psi  = reconstruct_state(result, N_basis=N_basis)
    xvec = np.linspace(xrange[0], xrange[1], pts)
    W    = qt.wigner(psi, xvec, xvec)
    return xvec, W

# ---- Internal helpers ----------------------------------------------------

def _fmt(v: float) -> str:
    """Replicate Julia's fmt(): %.10g with dots replaced by 'p'."""
    s = f"{float(v):.10g}"
    return s.replace(".", "p")
